File: crates/globstar/src/engine/ops/lower.rs
use crate::ast::Node;

use super::ir::{Op, OpProgram};

/// Lower an AST into one normalized [`OpProgram`]. Brace alternatives remain
/// nested, so ordinary in-segment braces never incur cartesian expansion.
pub fn lower(node: &Node, case_insensitive: bool) -> OpProgram {
    let mut ops = Vec::new();
    let mut needs_distribution = false;
    lower_into(node, &mut ops, case_insensitive, &mut needs_distribution);
    if needs_distribution {
        ops.clear();
        let distributed = distribute_seps(node.clone());
        lower_into(
            &distributed,
            &mut ops,
            case_insensitive,
            &mut needs_distribution,
        );
    }
    fold_globstars_inplace(&mut ops);
    apply_leading_seps_at_start(&mut ops);
    OpProgram::from_normalized(ops, case_insensitive)
}

fn lower_into(
    node: &Node,
    out: &mut Vec<Op>,
    case_insensitive: bool,
    needs_distribution: &mut bool,
) {
    match node {
        Node::Literal(bytes) => push_op(out, Op::Lit(bytes.clone())),
        Node::Separator => push_op(out, Op::Sep),
        Node::AnyChar => push_op(out, Op::AnyChar),
        Node::Star => push_op(out, Op::Star),
        Node::Globstar => push_op(out, Op::Globstar),
        Node::Class(class) => {
            let class = if case_insensitive {
                class.expanded_ascii_case_insensitive()
            } else {
                class.clone()
            };
            push_op(out, Op::Class(class));
        }
        Node::Concat(children) => {
            for (i, child) in children.iter().enumerate() {
                if let Node::Brace(branches) = child {
                    let prev_sep = i > 0 && matches!(children[i - 1], Node::Separator);
                    let next_sep = matches!(children.get(i + 1), Some(Node::Separator));
                    if (prev_sep && branches.iter().any(leads_globstar))
                        || (next_sep && branches.iter().any(trails_globstar))
                    {
                        *needs_distribution = true;
                    }
                }
                lower_into(child, out, case_insensitive, needs_distribution);
            }
        }
        Node::Brace(branches) => {
            let mut lowered = Vec::with_capacity(branches.len());
            for branch in branches {
                let mut branch_ops = Vec::new();
                lower_into(
                    branch,
                    &mut branch_ops,
                    case_insensitive,
                    needs_distribution,
                );
                fold_globstars_inplace(&mut branch_ops);
                lowered.push(branch_ops);
            }
            push_op(out, Op::Alternation(lowered));
        }
    }
}

fn push_op(out: &mut Vec<Op>, op: Op) {
    match op {
        Op::Star if matches!(out.last(), Some(Op::Star)) => {}
        Op::Lit(bytes) => {
            if let Some(Op::Lit(previous)) = out.last_mut() {
                previous.extend_from_slice(&bytes);
            } else {
                out.push(Op::Lit(bytes));
            }
        }
        other => out.push(other),
    }
}

fn leads_globstar(node: &Node) -> bool {
    match node {
        Node::Globstar => true,
        Node::Concat(children) => children.first().is_some_and(leads_globstar),
        Node::Brace(branches) => branches.iter().any(leads_globstar),
        _ => false,
    }
}

fn trails_globstar(node: &Node) -> bool {
    match node {
        Node::Globstar => true,
        Node::Concat(children) => children.last().is_some_and(trails_globstar),
        Node::Brace(branches) => branches.iter().any(trails_globstar),
        _ => false,
    }
}

fn apply_leading_seps_at_start(ops: &mut Vec<Op>) {
    match ops.first_mut() {
        Some(Op::OptSegmentsSlash) => ops.insert(0, Op::LeadingSeps),
        Some(Op::Alternation(branches)) => {
            for branch in branches {
                apply_leading_seps_at_start(branch);
            }
        }
        _ => {}
    }
}

fn distribute_seps(node: Node) -> Node {
    match node {
        Node::Concat(children) => {
            let mut out = Vec::with_capacity(children.len());
            let mut iter = children.into_iter().peekable();
            while let Some(child) = iter.next() {
                let Node::Brace(branches) = child else {
                    out.push(distribute_seps(child));
                    continue;
                };
                let prev_is_sep = matches!(out.last(), Some(Node::Separator));
                let prev_sep_owned_by_globstar =
                    prev_is_sep && matches!(out.iter().rev().nth(1), Some(Node::Globstar));
                let absorb_prev = prev_is_sep
                    && !prev_sep_owned_by_globstar
                    && branches.iter().any(leads_globstar);
                let absorb_next = matches!(iter.peek(), Some(Node::Separator))
                    && branches.iter().any(trails_globstar);
                if !absorb_prev && !absorb_next {
                    out.push(distribute_seps(Node::Brace(branches)));
                    continue;
                }
                if absorb_prev {
                    out.pop();
                }
                if absorb_next {
                    iter.next();
                }
                let branches = branches
                    .into_iter()
                    .map(|branch| {
                        let mut sequence = Vec::with_capacity(3);
                        if absorb_prev {
                            sequence.push(Node::Separator);
                        }
                        sequence.push(branch);
                        if absorb_next {
                            sequence.push(Node::Separator);
                        }
                        distribute_seps(Node::Concat(sequence))
                    })
                    .collect();
                out.push(Node::Brace(branches));
            }
            Node::Concat(out)
        }
        Node::Brace(branches) => Node::Brace(branches.into_iter().map(distribute_seps).collect()),
        other => other,
    }
}

fn fold_globstars_inplace(ops: &mut Vec<Op>) {
    let mut write = 0usize;
    let mut read = 0usize;
    while read < ops.len() {
        if matches!(ops[read], Op::Globstar) && matches!(ops.get(read + 1), Some(Op::Sep)) {
            ops[write] = Op::OptSegmentsSlash;
            read += 2;
            write += 1;
        } else {
            if read != write {
                ops.swap(write, read);
            }
            read += 1;
            write += 1;
        }
    }
    ops.truncate(write);

    for i in 0..ops.len().saturating_sub(1) {
        if matches!(ops[i], Op::Sep) && matches!(ops[i + 1], Op::OptSegmentsSlash) {
            ops[i] = Op::SepRun;
        }
    }

    write = 0;
    read = 0;
    while read < ops.len() {
        if matches!(ops[read], Op::Sep) && matches!(ops.get(read + 1), Some(Op::Globstar)) {
            ops[write] = Op::SlashAnything;
            read += 2;
            write += 1;
        } else {
            if read != write {
                ops.swap(write, read);
            }
            read += 1;
            write += 1;
        }
    }
    ops.truncate(write);

    for op in ops {
        if matches!(op, Op::Globstar) {
            *op = Op::GlobstarAny;
        }
    }
}

File: packages/globstar/src/engine/ops/lower.js
import {
  N_ANYCHAR,
  N_BRACE,
  N_CLASS,
  N_CONCAT,
  N_GLOBSTAR,
  N_LITERAL,
  N_SEPARATOR,
  N_STAR,
  brace,
  classExpandedAsciiCi,
  concat,
  sep,
} from "../../ast.js";
import { LiteralFacts } from "../facts.js";
import {
  ANYCHAR_OP,
  GLOBSTAR_OP,
  GSTAR_ANY_OP,
  LEADING_SEPS_OP,
  OP_ALTERNATION,
  OP_CLASS,
  OP_GLOBSTAR,
  OP_LIT,
  OP_OPT_SEGMENTS_SLASH,
  OP_SEP,
  OP_STAR,
  OSS_OP,
  SEP_OP,
  SEP_RUN_OP,
  SLASH_ANY_OP,
  STAR_OP,
} from "./ir.js";

export function lower(node, caseInsensitive) {
  const ops = [];
  const flag = { needsDistribution: false };
  lowerInto(node, ops, caseInsensitive, flag);
  if (flag.needsDistribution) {
    ops.length = 0;
    lowerInto(distributeSeps(node), ops, caseInsensitive, flag);
  }
  foldGlobstars(ops);
  applyLeadingSepsAtStart(ops);
  const ci = !!caseInsensitive;
  return { ops, facts: LiteralFacts.extract(ops, ci), caseInsensitive: ci };
}

function lowerInto(node, out, caseInsensitive, flag) {
  switch (node.tag) {
    case N_LITERAL:
      pushOp(out, { kind: OP_LIT, bytes: node.bytes });
      return;
    case N_SEPARATOR:
      pushOp(out, SEP_OP);
      return;
    case N_ANYCHAR:
      pushOp(out, ANYCHAR_OP);
      return;
    case N_STAR:
      pushOp(out, STAR_OP);
      return;
    case N_GLOBSTAR:
      pushOp(out, GLOBSTAR_OP);
      return;
    case N_CLASS:
      pushOp(out, { kind: OP_CLASS, cls: caseInsensitive ? classExpandedAsciiCi(node) : node });
      return;
    case N_CONCAT: {
      const children = node.children;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (child.tag === N_BRACE) {
          const prevSep = i > 0 && children[i - 1].tag === N_SEPARATOR;
          const nextSep = i + 1 < children.length && children[i + 1].tag === N_SEPARATOR;
          if (
            (prevSep && child.branches.some(leadsGlobstar)) ||
            (nextSep && child.branches.some(trailsGlobstar))
          ) {
            flag.needsDistribution = true;
          }
        }
        lowerInto(child, out, caseInsensitive, flag);
      }
      return;
    }
    case N_BRACE: {
      const branches = [];
      for (const branch of node.branches) {
        const branchOps = [];
        lowerInto(branch, branchOps, caseInsensitive, flag);
        foldGlobstars(branchOps);
        branches.push(branchOps);
      }
      pushOp(out, { kind: OP_ALTERNATION, branches });
      return;
    }
  }
}

function pushOp(out, op) {
  if (op.kind === OP_STAR && out.length > 0 && out[out.length - 1].kind === OP_STAR) return;
  if (op.kind === OP_LIT && out.length > 0 && out[out.length - 1].kind === OP_LIT) {
    const previous = out[out.length - 1];
    const merged = new Uint8Array(previous.bytes.length + op.bytes.length);
    merged.set(previous.bytes, 0);
    merged.set(op.bytes, previous.bytes.length);
    out[out.length - 1] = { kind: OP_LIT, bytes: merged };
  } else {
    out.push(op);
  }
}

function leadsGlobstar(node) {
  if (node.tag === N_GLOBSTAR) return true;
  if (node.tag === N_CONCAT) return node.children.length > 0 && leadsGlobstar(node.children[0]);
  return node.tag === N_BRACE && node.branches.some(leadsGlobstar);
}

function trailsGlobstar(node) {
  if (node.tag === N_GLOBSTAR) return true;
  if (node.tag === N_CONCAT) {
    return node.children.length > 0 && trailsGlobstar(node.children[node.children.length - 1]);
  }
  return node.tag === N_BRACE && node.branches.some(trailsGlobstar);
}

function applyLeadingSepsAtStart(ops) {
  applyLeadingSeps(ops, false);
}

// `tailIsOss`: the enclosing program continues with an OptSegmentsSlash
// (union factoring lifts shared trailing `**/` behind the alternation),
// so an empty branch's fork starts with `**/` and needs the §8.5
// LeadingSeps.
function applyLeadingSeps(ops, tailIsOss) {
  if (ops.length === 0) return;
  const nextIsOss = ops.length > 1 ? ops[1].kind === OP_OPT_SEGMENTS_SLASH : tailIsOss;
  const first = ops[0];
  if (first.kind === OP_OPT_SEGMENTS_SLASH) {
    ops.unshift(LEADING_SEPS_OP);
  } else if (first.kind === OP_ALTERNATION) {
    for (const branch of first.branches) {
      if (branch.length === 0) {
        if (nextIsOss) branch.push(LEADING_SEPS_OP);
      } else {
        applyLeadingSeps(branch, nextIsOss);
      }
    }
  }
}

function distributeSeps(node) {
  if (node.tag === N_CONCAT) {
    const out = [];
    const children = node.children;
    let i = 0;
    while (i < children.length) {
      const child = children[i];
      if (child.tag !== N_BRACE) {
        out.push(distributeSeps(child));
        i++;
        continue;
      }
      const prevIsSep = out.length > 0 && out[out.length - 1].tag === N_SEPARATOR;
      const prevSepOwnedByGlobstar =
        prevIsSep && out.length >= 2 && out[out.length - 2].tag === N_GLOBSTAR;
      const absorbPrev = prevIsSep && !prevSepOwnedByGlobstar && child.branches.some(leadsGlobstar);
      const absorbNext =
        i + 1 < children.length &&
        children[i + 1].tag === N_SEPARATOR &&
        child.branches.some(trailsGlobstar);
      if (!absorbPrev && !absorbNext) {
        out.push(distributeSeps(child));
        i++;
        continue;
      }
      if (absorbPrev) out.pop();
      const branches = child.branches.map((branch) => {
        const sequence = [];
        if (absorbPrev) sequence.push(sep());
        // Loop, not spread: ~64k children would overflow the argument limit.
        if (branch.tag === N_CONCAT) for (const child of branch.children) sequence.push(child);
        else sequence.push(branch);
        if (absorbNext) sequence.push(sep());
        return distributeSeps(concat(sequence));
      });
      out.push(brace(branches));
      i += absorbNext ? 2 : 1;
    }
    return concat(out);
  }
  if (node.tag === N_BRACE) return brace(node.branches.map(distributeSeps));
  return node;
}

function upgradeTrailingSeps(branches) {
  for (const branch of branches) {
    const last = branch.length > 0 ? branch[branch.length - 1] : undefined;
    if (last === undefined) continue;
    if (last.kind === OP_SEP) branch[branch.length - 1] = SEP_RUN_OP;
    else if (last.kind === OP_ALTERNATION) upgradeTrailingSeps(last.branches);
  }
}

function foldGlobstars(ops) {
  let write = 0;
  let read = 0;
  while (read < ops.length) {
    if (ops[read].kind === OP_GLOBSTAR && ops[read + 1]?.kind === OP_SEP) {
      ops[write++] = OSS_OP;
      read += 2;
    } else {
      if (read !== write) ops[write] = ops[read];
      read++;
      write++;
    }
  }
  ops.length = write;

  for (let i = 0; i + 1 < ops.length; i++) {
    if (ops[i + 1].kind !== OP_OPT_SEGMENTS_SLASH) continue;
    if (ops[i].kind === OP_SEP) ops[i] = SEP_RUN_OP;
    // A branch-trailing `/` before `**/` is `**`-adjacent too:
    // `{src/,lib/}**/x` ≡ `src/**/x` ∪ `lib/**/x` (§7.0 + §12.3).
    else if (ops[i].kind === OP_ALTERNATION) upgradeTrailingSeps(ops[i].branches);
  }

  write = 0;
  read = 0;
  while (read < ops.length) {
    if (ops[read].kind === OP_SEP && ops[read + 1]?.kind === OP_GLOBSTAR) {
      ops[write++] = SLASH_ANY_OP;
      read += 2;
    } else {
      if (read !== write) ops[write] = ops[read];
      read++;
      write++;
    }
  }
  ops.length = write;

  for (let i = 0; i < ops.length; i++) {
    if (ops[i].kind === OP_GLOBSTAR) ops[i] = GSTAR_ANY_OP;
  }
}
